# Day 1 — Hello, Python

> ⚠️ **Why this matters.** Python is the world's most-used programming language. NASA uses it, Netflix uses it, Instagram is built on it, and the AI revolution runs on it. You're about to write your first Python, and from this moment forward, every problem you've ever seen and shrugged at — "oh someone else can do that" — becomes something you can take a swing at.

[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/jakkzz/prince-curriculum/blob/main/phase-1-python-cli/lessons/01-hello-python.ipynb)

## What you'll do today

**Time:** 90 minutes.

By the end:

- [ ] You've run your first Python expressions in a notebook
- [ ] You can use the Python REPL
- [ ] You understand the difference between an expression and a statement
- [ ] You can read a Python error message
- [ ] You've written and run your first `.py` file with `uv run`

## The mental model

Python is a conversation, like the shell. You ask it to evaluate something. It evaluates. It tells you what it got.

```
you:    2 + 2
python: 4
```

That's the entire programming language. Everything else — variables, functions, classes, libraries — is just dressing on top of *evaluate this thing, give me back the answer*.

> 💡 **In the wild:** Every data scientist at every tech company spends a huge fraction of their day in Jupyter notebooks like this one. The flow — type, run, see, refine — is *the* iteration loop for exploratory programming. You're learning it on Day 1 because it's the right tool, not the toy tool.

## 1. Your first cell — run this

Click on the cell below. Press `Shift + Enter` (or click the ▶ button) to run it.

In [1]:
2 + 2

4

`Shift+Enter` = run this cell, move to next. `Cmd+Enter` (Mac) / `Ctrl+Enter` (Windows) = run, stay here.

Notebooks show the last expression's value automatically. No `print()` needed when you just want to see what something evaluates to.

## 2. Variables

A variable is a name attached to a value. The `=` is not "equals" — it's "this name now refers to this value."

In [2]:
age = 16
age

16

In [3]:
next_year = age + 1
next_year

17

> ⚠️ **One name per concept.** Don't reuse `age` to mean different things in different places. Read your future self's code: `age` should always mean age. If you have a new concept, give it a new name.

## 3. Strings — and the magic of f-strings

Strings are text, wrapped in quotes. Double or single, your choice — be consistent.

In [4]:
name = "Prince"
greeting = "Hello, " + name + "!"
greeting

'Hello, Prince!'

That `+` concatenation works but it's ugly. The modern way is an **f-string** — an `f` before the quote lets you embed expressions inside `{}`:

In [5]:
f"Hello, {name}! Next year you will be {age + 1}."

'Hello, Prince! Next year you will be 17.'

**Try it:** Change the cell above to print your real name and your real age. Re-run.

## 4. Expression vs statement

This trips up beginners. Memorize it now.

- **Expression** = code that *produces a value*. `2 + 2`, `len("hello")`, `name.upper()`.
- **Statement** = code that *does something*. `name = "Prince"`, `if x: ...`, `print(x)`.

An expression can be on the right side of `=`. A statement cannot.

Notebooks show the value of the **last expression** in a cell. They don't show anything for statements unless you explicitly print.

In [6]:
x = 5         # statement — assigns 5 to x, no output

In [7]:
x             # expression — produces 5

5

In [8]:
print(x)      # statement — does something (prints) but has no output value

5


Notice the difference: the bare `x` cell showed `5` as the *result* (in `Out[]`). The `print(x)` cell showed `5` as *output stream* (no `Out[]`).

It's subtle but it matters when you start writing functions.

## 5. Numbers — int vs float

Python has two kinds of numbers:

- `int` — whole numbers. No size limit. `42`, `-7`, `1_000_000` (underscores for readability).
- `float` — decimals. IEEE 754, so imprecise. `3.14`, `-0.5`, `1.0`.

**Math operators:**

| Operator | Meaning | Example |
|----------|---------|---------|
| `+`      | add | `2 + 3` → `5` |
| `-`      | subtract | `7 - 4` → `3` |
| `*`      | multiply | `2 * 3` → `6` |
| `/`      | divide (always float) | `7 / 2` → `3.5` |
| `//`     | floor divide (integer result) | `7 // 2` → `3` |
| `%`      | modulo (remainder) | `7 % 2` → `1` |
| `**`     | exponent | `2 ** 10` → `1024` |

Run each below:

In [9]:
7 / 2     # always returns a float, even when result is whole

3.5

In [10]:
7 // 2    # floor division — drops the remainder

3

In [11]:
0.1 + 0.2    # this is not 0.3 — welcome to floating point

0.30000000000000004

> ⚠️ **Float weirdness is real.** `0.1 + 0.2 != 0.3`. This isn't a Python bug — it's how computers represent decimals in binary. When you need exact decimals (money!), use the `Decimal` type. For everything else, just be aware.

## 6. Reading error messages

You will get errors. Reading the error is half the battle. Most beginners scan past them; senior engineers read them carefully.

**Run the cell below.** It will error on purpose. Don't panic — read the message.

In [12]:
5 + "hello"

TypeError: unsupported operand type(s) for +: 'int' and 'str'

Read the error from bottom to top:

1. **The bottom line is the actual error:** `TypeError: unsupported operand type(s) for +: 'int' and 'str'`. Translation: "I don't know how to `+` a number and a text."
2. **The line above shows where:** Cell In[12], line 1.
3. **The arrow `--->`** points to the offending line.

Errors are the language being **honest** with you. They tell you what's wrong, where, and often hint at how to fix it. The fix here is: convert the number to a string first.

**Try it:** Modify the cell to be `str(5) + " hello"`. Re-run. Now it works.

## 7. From notebook to script

Notebooks are great for exploration. Real programs live in `.py` files.

**Outside this notebook**, in your terminal:

```bash
cd ~/prince/scratch/python-warmup
cat > greet.py << 'EOF'
name = input("Your name: ")
age = int(input("Your age: "))
print(f"Hi {name}! Next year you'll be {age + 1}.")
EOF

uv run python greet.py
```

Expected output:

```
Your name: Prince
Your age: 16
Hi Prince! Next year you'll be 17.
```

**Same Python.** Different file. The notebook is for trying things. The script is for shipping things.


## Mini-exercise — build a tiny temperature converter

In a new code cell below, write Python that:

1. Sets a variable `celsius` to some integer (your choice, e.g. `28`).
2. Computes the Fahrenheit equivalent using the formula `F = C × 9/5 + 32`.
3. Stores it in a variable `fahrenheit`.
4. Prints `"28°C = 82.4°F"` — using an f-string, with both values embedded.

Expected output (if you start with `celsius = 28`):

```
28°C = 82.4°F
```

Try it before peeking at the answer below.

In [ ]:
# Your code here


<details>
<summary>Solution (try first!)</summary>

```python
celsius = 28
fahrenheit = celsius * 9 / 5 + 32
print(f"{celsius}°C = {fahrenheit}°F")
```
</details>

## Connect to the project

> 🎯 **Connects to the project:** Your Phase 1 capstone `pomo` will use exactly these building blocks — f-strings for display (`f"Started session #{id} — {duration} min"`), input/output, `int()` and `str()` conversions for parsing user input. The Python you're writing in this notebook is **the same Python** that goes into your real tool. Today's playground is tomorrow's production code.

## Self-check

<details>
<summary>1. What does <code>=</code> mean in Python?</summary>

Assignment, not equality. `age = 16` means "the name `age` now refers to the value `16`." Test equality with `==`.
</details>

<details>
<summary>2. What's the difference between an expression and a statement?</summary>

An expression *evaluates* to a value (`2 + 2`, `name`, `f"hi {name}"`). A statement *does* something (`name = "Prince"`, `print(x)`, `if x: ...`). Expressions can be on the right of `=`; statements can't.
</details>

<details>
<summary>3. Why does <code>0.1 + 0.2</code> not equal <code>0.3</code> in Python?</summary>

Floats are stored in binary IEEE 754 format. Some decimals don't have exact binary representations. `0.1 + 0.2` is actually `0.30000000000000004`. For money or exact decimals, use `decimal.Decimal`.
</details>

<details>
<summary>4. What does this error mean? <code>TypeError: unsupported operand type(s) for +: 'int' and 'str'</code></summary>

You tried to `+` a number and a string. Python won't auto-convert. Wrap the number with `str()` or the string with `int()`, whichever makes sense.
</details>

<details>
<summary>5. How is a Jupyter notebook different from a <code>.py</code> file?</summary>

Notebooks are interactive — cells run one at a time, state persists between cells, results display inline. Great for exploration. `.py` files are scripts — run top to bottom, no inline output, deploys to production. You'll use both: notebooks to try things, scripts to ship them.
</details>

## What's next

Tomorrow: **types and values** in depth. Why Python has so many "kinds of things," what they're each good for, and the type hints that make your code self-documenting.